### ISSUE TITLE
Constant Stutter from TV4Play

### ISSUE THREAD

**Issue by RickardNi** [spaam/svtplay-dl#1046]
### What is the issue:

When downloading clips from TV4Play, the video is constantly stuttering which is very noticable, especially on TV or larger screens. The issue does not occur when playing in the browser, and was not present on earlier versions for the same clips.

Example link:
https://www.tv4play.se/program/s%C3%A5-mycket-b%C3%A4ttre/3971905

Tried it in November with 2.0 when I noticed the issue, and now with 2.1 and months later, the issue still persists. Any ideas? It can very well be ffmpeg or library issues, I guess.

**Comment by Sopor-** [comment]
I wondering if this can be the same as #982?

**Comment by RickardNi** [comment]
It might be, I haven't played the video for long, since it's unbearable to watch.

However, I did some testing now and copied the .ts file before it was complete (and muxed), and then the problem didn't occur. After merging the video and audio file, the problem occured again.

I wonder why this is only present on TV4Play, but you know more than I do. That also explains why it occurs on later versions and not on earlier versions because back then I did the mux manually with ffmpeg.

I tried to find an option to turn off the muxing, but couldn't find one. If I could just be allowed to download them separately and mux them myself, I think everything would be fine.

**Comment by qnorsten** [comment]
which ffmpeg version are you running? 
which command do you manually run to mix it?

**Comment by Sopor-** [comment]
> However, I did some testing now and copied the .ts file before it was complete (and muxed), and then the problem didn't occur. After merging the video and audio file, the problem occured again.

@RickardNi It sounds exactly as my issue. When i merge video and audio i get the issue.

**Comment by RickardNi** [comment]
@Sopor- But I get the issue with 4.0.2 which should be 4.x or higher? Will try to update version and see if it makes a difference. I know I updated to the latest at the time when I had the problem initially and some weeks after that.

**Comment by RickardNi** [comment]
Tried with ffmpeg 4.1 now and the same issue occurs.

**Comment by Sopor-** [comment]
Yeah, i know. It is only the sync issue you will fix by updating ffmpeg to 4.x. At least i found someone with the same issue that i have on TV4. So now i'm not alone :)

**Comment by RickardNi** [comment]
Ah, sorry, I misunderstood. Yeah, it's really annoying, but at least easily produced, and very noticable in the video I linked to (after the jingle-introduction-whatever-it's-called).

**Comment by Sopor-** [comment]
@RickardNi Could this be the reason for the bad playback?
```
[mpegts @ 00000029f2259500] Invalid DTS: 239274000 PTS: 239270400 in output stream 0:0, replacing by guess
[mpegts @ 00000029f221ae40] Invalid timestamps stream=0, pts=240199200, dts=240202800, size=4061
[mpegts @ 00000029f2259500] Invalid DTS: 239284800 PTS: 239281200 in output stream 0:0, replacing by guess
[mpegts @ 00000029f2259500] Invalid DTS: 239292000 PTS: 239288400 in output stream 0:0, replacing by guess
[mpegts @ 00000029f2259500] Invalid DTS: 239302800 PTS: 239299200 in output stream 0:0, replacing by guess
```
I get these error messages when i merge the audio and video manually.

**Comment by Sopor-** [comment]
It seems that the issue is the audio file and if i merge the video and audio with `-fflags +igndts` i get a smooth playback on TV4. Maybe @spaam could take a look at it?
It seems that the timestamps are broken and this will regenerate DTS based on PTS.

**Comment by granbom** [comment]
@Sopor-and @RickardNi 
I agree, there are issues with tv4 files.
My problem last 5-6 months"ish"  have been that the sound is lagging during playback maybe 0.1 second every 3-4 minutes. So after a while even GW doesn't make sence...
No "studdering" in the video though, but you have to delay the audio in the playback 100 ms every couple of minutes.
This issue I've only seen with files from the above provider...

**Comment by Sopor-** [comment]
@granbom What media player are you using? Maybe some of them make the audio out of sync (the issue seems to be the audio) and other makes the video bad? I'm using Kodi, VLC and Media Player Classic Homecinema and they all make a bad video. Kodi also seems to stop a couple of time during playback so i have to resume, but i have never seen out of sync if i use ffmpeg 4.x or higher.

**Comment by granbom** [comment]
@Sopor- I'm using VLC. I use ffmpeg 4.1 (from brew on macOS 10.13) usually on Ubuntu I compile ffmpeg my self, then the version number is the commit number from git, but yes, above 4, but I guess the lib* version numbers are the ones to care about in ffmpeg.
I did now however notice that yesterday tv4 had their max HLS quality for latest "brottsjournalen" around 1800, and I got lagging audio. Today they have more bitstreams, highest 4462, and that one was fine. Also, their current HLS 1926 (1280x720) one is also fine. Could be that it's actually them who have issues. The lagging isn't notices if you watch on tv4play because the commercial breaks reset the audio lag. (really "killgissar" now, but a thought)

Also, if I compare these two files now. The one from yesterday (1677 kb/s, 1280x720) is 44:47.84 long (checking with ffprobe) and the one from today is 44:06.93 (1825 kb/s, 1280x720) 
Hence, tv4 have made changes since yesterday and the lag is gone.
Their 4218 kb/s, 1920x1080 file is also 44:06.93 and fine. That one didn't exist yesterday morning.

**Comment by Sopor-** [comment]
I have this video lag in almost every programs i download from TV4, Brottsjournalen, Maria Wern, Parlamentet, STHLM Rekviem and i always download highest bitrate and i use `-P hls`. I did noticed that `Haffad på gränsen` is playing without any lag or stutter.
@granbom I assume you have at least 100 mbit so i could send you two videos? One that have this lag and another without (the one i made with `-fflags +igndts`). Then you could see if you have the same lag and stutter as i have... Let me know if you are interested...

**Comment by granbom** [comment]
@Sopor- sure, where can I download the files?

**Comment by Sopor-** [comment]
@granbom You will notice it at the beginning when they zoom in to Jenny and of course in the whole video. [good](https://ufile.io/lie0q) and [bad](https://ufile.io/wu8r2)

**Comment by granbom** [comment]
@Sopor- Here are my findings and guessing. I also downloaded episode 1 of brottsjournalen, but changed svtplay-dl so it didn't merge them, hence one video ts and one audio ts and I got the same result as your bad.mp4 and good.mp4.
In the S02E01 file I see a lot of:
`[mpegts @ 0x55b998ee5180] Invalid timestamps stream=0, pts=240163200, dts=240166800, size=3588`
errors.
I also downloaded S02E04 and these errors are not in episode 4.

For e01 I do the same as you have done with +igndts and that produces a good file. If not using igndts that is the same result as the bad file. No issues for e04 (and e04 doesn't have any invalid timestamps errors either).

**good**
```
ffmpeg -fflags +igndts -i brottsjournalen.s02e01.ts -i brottsjournalen.s02e01.audio.ts -c copy -bsf:a aac_adtstoasc brottsjournalen.s02e01.igndts.mp4
```
**bad**
```
ffmpeg -i brottsjournalen.s02e01.ts -i brottsjournalen.s02e01.audio.ts -c copy -bsf:a aac_adtstoasc brottsjournalen.s02e01.igndts.mp4
```
As I understand it, +igndts is for fixing broken ts streams, so it's not weird it produces a good result.

Why is it broken then?
After a bit of googling, it could happen when using HLS and a source mp4 file, that is then converted to ts during streaming and the conversion doesn't work as intended.

So TV4 produced an OK file for E04, but a broken one for E01 (that can be fixed with +igndts). Using +igndts for all files is not recommended as I understand it, and can have side effects if the ts file is not broken.

**Comment by granbom** [comment]
I'm not sure, but I guess the changes to svtplay-dl could be:

1. Do a test muxing without +igndts for like 5 seconds (like `ffmpeg -i input.ts -i input.audio.ts -c copy -ss 00:01:00 -t 00:00:05 -f null -`)
2. Read the ffmpeg stderr and look for `Invalid timestamps` messages (in S02E01 there are thousands  of them)
3. If so, add -fflags +igndts to the muxing parameters, othervise mux without them.
